### PR TITLE
[SPARK-8591][CORE]Block failed to unroll to memory should not be repl…

### PR DESCRIPTION
**Pull request by dibbhatt**
Block failed to unroll to memory should not be replicated for MEMORY_ONLY_2 StorageLevel

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 1]
```diff
@@ -834,7 +837,9 @@ private[spark] class BlockManager(
 
     // Either we're storing bytes and we asynchronously started replication, or we're storing
     // values and need to serialize and replicate them now:
-    if (putLevel.replication > 1) {
+    // Should not replicate the block if putBlockStatus is StorageLevel.NONE
+    if (putBlockStatus != null &&
+      putBlockStatus.storageLevel != StorageLevel.NONE && putLevel.replication > 1) {
```
Comment (lianhuiwang) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: i think we can add isFailed function to BlockInfo. and this line can be changed to following:
if (putBlockInfo.isFailed && putLevel.replication > 1) {}

**Comment by andrewor14** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 1 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: +1
I like limiting the side effects to `putBlockInfo`.  (though tiny nit, I think you forgot a `!`, should be

 `if (!putBlockInfo.isFailed && ...`

Comment (lianhuiwang) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: yes, thanks.

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Thanks for your review. I will make the changes.

**Comment by sujkh85** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-8591][CORE]Block failed to unroll to memory should not be repl… (#6990)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by dibbhatt** [comment]
Hi @andrewor14 @lianhuiwang @squito let me know if latest changes looks fine

File: core/src/main/scala/org/apache/spark/storage/BlockInfo.scala [hunk 2]
```diff
@@ -70,6 +70,11 @@ private[storage] class BlockInfo(val level: StorageLevel, val tellMaster: Boolea
       this.notifyAll()
     }
   }
+
+  /** Check is this BlockInfo is failed */
```
Comment (lianhuiwang) [inline core/src/main/scala/org/apache/spark/storage/BlockInfo.scala]: may be Check whether this BockInfo is failed

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 3]
```diff
@@ -834,7 +834,8 @@ private[spark] class BlockManager(
 
     // Either we're storing bytes and we asynchronously started replication, or we're storing
     // values and need to serialize and replicate them now:
-    if (putLevel.replication > 1) {
+    // Should not replicate the block if putBlockStatus is StorageLevel.NONE
```
Comment (lianhuiwang) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: // Should not replicate the block if its StorageLevel is StorageLevel.NONE or putting it to local is failed
is better.

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: done

File: core/src/main/scala/org/apache/spark/storage/BlockInfo.scala (hunk 2 above)
Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockInfo.scala]: Done

File: core/src/main/scala/org/apache/spark/storage/BlockInfo.scala [hunk 4]
```diff
@@ -70,6 +70,11 @@ private[storage] class BlockInfo(val level: StorageLevel, val tellMaster: Boolea
       this.notifyAll()
     }
   }
+
+  /** Check whether this BockInfo is failed */
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockInfo.scala]: typo: BlockInfo (missing 'l')

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 5]
```diff
@@ -833,8 +833,10 @@ private[spark] class BlockManager(
     logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
 
     // Either we're storing bytes and we asynchronously started replication, or we're storing
-    // values and need to serialize and replicate them now:
-    if (putLevel.replication > 1) {
+    // values and need to serialize and replicate them now.
+    // Should not replicate the block if its StorageLevel is StorageLevel.NONE or
+    // putting it to local is failed.
+    if (!putBlockInfo.isFailed && putLevel.replication > 1) {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I think the comment about `StorageLevel` is no longer relevant, right?  Also I prefer the extended comment explained *why* this is necessary, eg. something along the lines of:

Replicating requires enough memory for the whole serialized block.  But if we failed to put the block locally, that means we didn't have enough memory, so we shouldn't try to replicate.

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: hi @squito @lianhuiwang,  let me know how this comment sounds ... 

"if a block is not serialized to memory locally that means we don't have enough memory. In this case we should not try to replicate the same block to remote peer even though remote peer has enough memory to serialize the block . BlockInfo status 'failed' for local block indicates that block failed to serialize locally and we can skip replicating the same block."

Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: blocks aren't necessarily *serialized* to memory locally, they are just stored (they might be serialized, or they might not).  I think this version is still missing the *why* component a little bit -- that if you don't have enough memory to store locally, you don't even have enough memory to create the message that requests the peer to replicate.

Comment (lianhuiwang) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: yes,i agree with StorageLevel is no longer relevant. maybe we need to think more about why block is failed.because as @squito said it might be not serialized,it just put block to entries and does not have more memory. right?

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Agree with this. BlockInfo can be failed for MEMORY_ONLY ( or MEMORY_ONLY_2) cases when Blocks not able to Unroll Safely (In case of IteratorValue)  or not able to store to memory for ArrayValue irrespective of Block is serialized or not. Thus if Block is not stored , or in other words , if Block is either not Unrolled Safely or Try to Put the block failed because there is not enough memory , we should not do Replicate the Block .

But I am not able to understand what @squito said "you don't even have enough memory to create the message that requests the peer to replicate" . 

If you see the test cases , if Block Unroll Failed , it can still Replicate to Peer if BlockInfo Failed check is not done (as it was before the fix). I am trying to prevent the failed block to replicate because store to local has failed and block status is not reported to master.

Comment (lianhuiwang) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: yes, now i understand. when Block is received but memoryStore does not have enough memory, so Block is Failed in local. block's memory have been requested before putting to memoryStore and then memoryStore will check whether number of its memory has enough than block size. number of its memory is from memoryFraction, but now JVM maybe have enough memory and can create any object.

Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: oops, I thought the purpose of this patch was something else earlier.

Sorry to make you explain this again, but I am not certain I understand what is wrong w/ the current behavior.  If we fail to put locally, with the current behavior, (1) the block will get replicated anyway (2) the remote peer will cache the block in the originally requested storage level if it can (eg., it will store the block in memory) since [the replicate request uses the original `putLevel`](https://github.com/apache/spark/blob/a56516fc9280724db8fdef8e7d109ed7e28e427d/core/src/main/scala/org/apache/spark/storage/BlockManager.scala#L853) and (3) the remote node will tell `BlockManagerMaster` that it has the block, b/c [`putBlockData`](https://github.com/apache/spark/blob/a56516fc9280724db8fdef8e7d109ed7e28e427d/core/src/main/scala/org/apache/spark/storage/BlockManager.scala#L316) just calls [`putBytes`](https://github.com/apache/spark/blob/a56516fc9280724db8fdef8e7d109ed7e28e427d/core/src/main/scala/org/apache/spark/storage/BlockManager.scala#L676) which defaults to `tellMaster = true`.

Is my explanation of the current behavior incorrect, or is there something wrong w/ the current behavior that I'm not seeing?

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: hi @squito,  yes this explanation is correct. But in current behavior the desired replication level '2' will not be maintained if Block failed in Local but replicated anyway , as Block is now will be in one Node even though the Replication factor is 2 .

Comment (lianhuiwang) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: i think @squito said is correct. now the problem is when block is failed in local and replication factor needs 2, but now how we choose to deal with this situation, there are three ways: 
one is making replication factor to 1 and continue to replication to remote node. 
two is stop to replication to remote node and its replication factor is 0. 
three is continuing to replication to two remote node and its replication factor is 2.
so i think the three maybe a good solution. how about your opinions? @andrewor14 @squito

Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Thanks for the clarification.  Yes, I think you are right about the 3 options @lianhuiwang .  I'm sorry I didn't understand what was going on earlier, and I feel bad for saying this b/c I know this has already been through a few rounds of revision -- but I'm not sure I see the point in implementing option 2, as this patch does in its current form.  It doesn't seem particularly bad to me if you request replication level 2 in memory, and you only end up with 1 replicate in memory.  In fact, this can happen a few other ways -- what if you do manage to store locally, but no peer has enough memory?   Or what if you get 2 replicates stored now, but later on only one copy gets evicted from the cache so you are down to one copy in memory?

(All that being said, I do see the appeal of this patch as a *heuristic* fix -- if you can't store the data locally, you are unlikely to be able to store the data on a remote, so maybe you shouldn't bother?  But if that is the reason we are making this change, that is exactly what I'd like the comment here to indicate.)

What about option 1 (current behavior -- just 1 replicate) & option 3 (change behavior -- replicate to two nodes). Option 3 is slightly appealing ... but for consistency, it would also suggest that if you use `StorageLevel.MEMORY_ONLY` (ie. only 1 replicate), and you can't store locally, that then you would also try to find a remote node to store the block on.   And I really doubt we want to do that -- that is a pretty big change in behavior for how `StorageLevel.MEMORY_ONLY` works.

I'm really on the fence on this.  Lets see what others think.

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I agree with you @squito . What I have observed while working with Spark Streaming is , there is a higher probability of getting the BlockNotFoundException if I use MEMORY_ONLY settings and blocks got evicted from memory. The MEMORY_ONLY_2 can help that to some extent as even if block get dropped form one node, it still be available to other. This fix will help to reduce the BlockNotFoundException while storing the Blocks to local ( and failed ) , thus the Receiver gets the exception and can re-try again to store at some point in future . If we allow block to replicate even if store to local is failed, which is going to reduce the desired replication factor , Receiver will not have any option to re-try . In case of Local store to success and remote failed , or both local and remote store is success but eventually at some point block got evicted is not at receiver's control. So out of three possible cases ( Block failed locally but stored remotely , block stored locally but failed remotely , block stored both in remote and local but evicted sometime in future ) , this fix can solve the BlockNotFoundException for 1/3 rd of the cases by giving the Receiver to re-try to store the block once again to get desired replication factor 2. 

Do let me know if this argument is valid ?

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Also as I see the replicate process, if one remote peer is failed to replicate , it tries with other remote peer to get desired Replication factor 2, so the retry mechanism to replicate block already in place.  Thus if store to local is failed and if we stop replication to happen, this gives higher chance for client to retry same block and get desired replication factor .

Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: ah, ok I hadn't considered streaming before.  I'm not an expert on that part of the code, so maybe we should rope in somebody else more knowledgeable -- @tdas @harishreedharan , want to weigh in here?

With that caveat, I think I see the issue now.  If I understand correctly, the problem is that if you can't store a block locally, the receiver thinks the block has not been stored anywhere -- even if it has been successfully replicated.  That is because the receiver [just looks at the result of `doPut`](https://github.com/apache/spark/blob/3a342dedc04799948bf6da69843bd1a91202ffe5/streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceivedBlockHandler.scala#L91) to see if the block has been stored.  But, as you've noted, when we fail to store locally but successfully store on a remote, that result doesn't contain the block.

So your proposal seems to be -- as long as the receiver is going to ignore the replicated block in any case, there isn't any sense in replicating.  But maybe a better alternative would be for the receiver to accept it being stored on a remote?

This makes me wonder if the receiver should also be checking that the block has actually been replicated, and not only stored locally?

Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: i can see that its beneficial to throw errors if replicating to two was not possible, so that the receiver can retry. However, even if the receiver retries, there is no good way for the receiver to ensure that the block has been replicated to the desired level even after two tries. Since there is not feedback mechanism to check for success after retries, doing something that increases "likelihood" is not very useful. Doing something like that and relying on is bad design. 

That's why I am more inclined towards option 3, that is, if local fails, it tries to replicate it two machines. I agree that it is inconsistent with MEMORY_ONLY, but its still better change than the above which does not provided anything significantly more. And I think its okay to break consistency because of the benefit we are getting especially in a scenario where the behavior is related to something as critical as fault-tolerance behavior.

Furthermore, stepping back, for receivers, why are even using MEMORY_ONLY and not MEMORY_AND_DISK (with or w/o replication)? Do you get any benefit by using the former over latter?

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Right , if Block failed to stored locally , the Receiver will not get the BlockId as it looks for the BlockId in PutResult even though block got replicated . But @tdas and @harishreedharan can confirm if this indeed the issue. My point is even if Receiver can check if Blocks are replicated in this case, the Replication factor is already down to 1 and thus have higher chance of getting BlockNotFoundException and streaming jobs failed .

Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: @dibbhatt I think you missed reading my comment when you posted. Race condition we had :)

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Hi @tdas it was indeed race condition. I did not see your reply before posting mine :)

I understand your point completely that options 3 is better , and it is also true that MEMORY_AND_DISK cases wont be having this block dropping issue . But if someone use MEMORY_ONLY_2 settings , then there will be a possible problem.

Nevertheless , my point here was little different . There are two possible cases being discussed here.. . 

Case 1 : Even in present state , for MEMORY_ONLY_2 Storagelevel, if Block failed to stored locally , but replicated to remote peer , the ReceivedBlockHandler wont be getting the block id and will throw Exception . Now it is upto the Receiver implementation to ignore this exception or retry depending on how resilient the Receiver is towards data loss. My point is , if Receiver anyway not getting the Block Id if BlockManager failed to store locally, why even then try to Replicate ? It unnecessary consume additional memory in remote peer. And I guess it can happen to any client which calls BlockManager doPut to store the block and get updatedBlocks object which is Seq[(BlockId, BlockStatus)] to check if block id exists . As this updatedBlocks  does not contains the details if replication is successful or not , client wont find the blockId if store to local is failed. 

Case 2 : What @squito is telling is also correct. Even if block failed to store locally but replicated to remote peer, why cant ReceivedBlockHandler get the block id from replicated node before throwing Exception. But that need change in ReceivedBlockHanlder to get block details from remote peers. But in this case the desired replication level will be reduced to 1 instead 2 .

If we try option 3 , even in that case , I guess ReceivedBlockHandler needs changes to get the block id from remote peer.

Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Case 2: Getting the block status from the remote peer is definitely not the
goal of a block manager and is not the intended semantics of PutResult

Case 1: I think I get your point. For the receiver point of view, local
store has failed, why even replicated. However, from the point of the block
manager which manages stuff for both receivers as well as RDD partitions,
this is definitely tricky to change the semantics right now. For RDD
partitions, this change would be a regression as the replica used to still
help speed up stuff.

Let me first confirm. Is the problem that you are trying to solve that
there are received blocks that gets inserted but never cleared?

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: The problem here is , if local memory got filled up and block store failed, blocks still get replicated to remote and used up memory but same blocks never used in Streaming jobs... Even though those blocks will eventually evicted , but this fix will optimize the memory.  I understand your concern about RDD partition which can still use the remote replica for speedup even local store failed.

Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Since those blocks are never used and eventually gets evicted, the only
problem is slightly higher memory usage for some time. So I dont really see
a critical problem that needs to be solved at the cost of the de-optimizing
existing code paths.

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: thanks @tdas for your view. I agree with what you said . Just for my understanding , for RDD Partitions which are cached can also be evicted from memory by BlockManager which is more often the case in Spark . But for Spark Streaming case , the blocks are generating at much faster rate and cause large number of unwanted blocks to be stored in remote memory if local store failed. If Receiver do retry those blocks (which normally the case ) and retry is success , there will be multiple copies of same blocks will exists across nodes. This seems to me little unwanted .
Not sure why you called de-optimizing existing code path, as for RDD partitions to be dropped from memory can happen anytime during its life cycle.

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Also as I see the code path of doing rdd.Cache , the CacheManager first check if block can be unrolled safely , and then only it tries to call blockManager.putArray(..) which returns the updatedBlocks object. If block is not unrolled safely , CacheManager wont cache the block. So I do not think PR will de-optimize existing code path , as in RDD partitions case, once block is unrolled safely , CacheManager store the block and get the updatedBlocks back, and in this case my fix wont impact anything and replication will happen if block is unrolled safely.  @squito do let me know if this is correct.

Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: @dibbhatt I think I see what you're getting at, but I'm with TD that I don't know if this change is worth it.  I believe the "de-optimizing" he is referring to is that in normal spark, if you store remote but not local, it is useful.  Normal spark jobs will take advantage of that remote copy.

This change is really only designed to help out with streaming, where the recommendation is to use MEMORY_AND_DISK anyway.  IMO its not worth any of the changes for that -- even option (3) would introduce a decent amount of complexity in the BlockManager that would require a big comment to explain why we do that ... and it would just be to support this one use case that isn't recommended anyway.

Just to brainstorm, I want to throw one more idea out there.  We don't want the receiver to have to do a query after `doPut` to see if the block was replicated somewhere, b/c that would be too expensive.  But what if instead we change the return type of `doPut` to already include that information?  `doPut` already knows if replication was successful, so instead of returning `Seq[(BlockId, BlockStatus)]`, it could return `case class DoPutResult(updatedBlocks: Seq[(BlockId, BlockStatus)], replicationLevel: Int)` (probably with some better name).  Then the receiver would have the information it needs without needing to perform any expensive extra checks.

That said, I'm not really favor of this idea either -- again I feel its a decent amount of complexity, without a compelling reason.  thoughts @tdas?

sorry @dibbhatt but I'm leaning towards "won't fix / not a bug" for this issue.

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: not a problem @squito ..:) . If TD does not have any further comments , we can close this Jira .

Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: @squito Your new idea also involves quite invasive changes, I am not sure the use case is strong enough to warrant that. Here is a different idea. How about doing the unrolling in the same fashion as the RDD cache? Unroll first into an array, and then put it in the block manager or throw exception. How about that?

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: hi @tdas , the implementation of BlockManager put* (putArray, putIterator)  methods exactly does the same. it calls the MemoryStore unrollSafely , and if block is unrolled safely , then only it cache the block . What this PR does is , if block not unrolled to memory, and storage level is not using Disk , it stops the block to replicate , and at the same time the updatedBlock result will not contain the blockId which leads to ReceivedBlockHandler throw the exception. As I said this fix wont impact the CacheManager flow , as in CacheManager it explicitly does the same flow ( check for unroll safely first and then cache the block). 

Not sure why you think this is not a strong use case, as I see for high volume Receivers ( say consuming from large partitions of Kafka topic ) receiving from say 50+ receivers from hundreds of partitions , will unnecessarily trigger the Replication process even if block can not stored in local memory and those block will never going to be used in Streaming jobs. This will leads to Receivers which running on Remote peer will also to face memory issue as lot of unnecessary blocks clogging its memory..So I see this leads to chaining event which impact every Receivers ..

Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I will let @andrewor14 chime in on this, to figure out what is best
solution here for Spark Core - change block manager, or not. Accordingly
this PR should change block manager, or change ReceivedBlockHandler.

Comment (dibbhatt) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Hi @andrewor14 , kindly let me know your thought on this .

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Hi, sorry on slipping. I haven't been following this discussion but I'll do so today.

**Comment by andrewor14** [comment]
@dibbhatt Thanks for submitting the fix. I spoke offline with @tdas about this issue and came to the following conclusions.

**Problem summary.** The issue seems to be a disconnect between what the `BlockManagerMaster` knows and what the `ReceiverTracker` knows. If a block fails to unroll, the receiver tracker will never know about the block and will not include it in a future computation. In the mean time, however, the block may be replicated and take up space on other executors even though it will never be used.

**Implications for Spark core.** For Spark core, however, it is reasonable to replicate a block even if it fails to unroll. Just because there is not enough memory to cache this block on this executor doesn't mean the same is true on a different executor. This is all best effort, but a future computation of the RDD will still benefit from having the cached block somewhere. (Note: the existing code doesn't actually do this yet because `CacheManager` has its own unrolling logic. But replication even when unrolling fails is something we want in the future for normal RDD caching.)

**Alternative fix.** The right fix in Spark streaming would be to have the `ReceiverTracker` just read its blocks from the `BlockManagerMaster`. This simplifies the two divergent block reporting code paths. Since the `BlockManagerMaster` is notified of replicated blocks, the replication here will also help mitigate data loss in the case of `MEMORY_ONLY_*`.

**TL;DR.** This patch removes a small feature from block manager that, though not currently used, is desirable in the future. However, the underlying issue is not caused by a bug in the block manager, but an incorrect assumption in the `ReceiverTracker` that doesn't take into account replication. The correct way forward would be to fix this in Spark streaming by refactoring the `ReceiverTracker` to depend on `BlockManagerMaster`.

In short, we should consider an alternative fix that is longer-term and more correct.

**Comment by dibbhatt** [comment]
@andrewor14 thanks for summarizing the problem and how to solve this. I will go ahead and close this PR. 

@tdas , shall I raise a separate JIRA to track how ReceiverTracker can get the block id from BlockManagerMaster ?

**Comment by tdas** [comment]
@dibbhatt 

Yes, you can close this PR.

Regarding the alternative fix of finding the blocks from the BlockManagerMaster, I thought about it a little bit more, and its seems non-trivial to make that change. Right now the direct route of block reporting [ReceiverBlockHandler <--> ReceiverTracker] ensures that by the time this message exchange completes the block has been received reliably, especially in case of WAL and stuff. Its easy to reason about the fault-tolerance guarantees. If the report is to be reimplemented indirectly through the BlockManager, its non-trivial to changes things such that the guarantee is maintained. So I am not super inclined to make make a new JIRA for that. 

Rather what I am inclined to do is to modify this existing JIRA to flag this issue specifically for Spark Streaming (so that people can find it), and then close it as Won't Fix for now. This is not a big issue as - 1. people really should not be doing obviously problematic thing as use MEMORY_ONLY
2 even if it occurs, the extra blocks will just drop out of memory eventually anyways as it wont be used and LRU will push out those blocks.

**Comment by dibbhatt** [comment]
Thanks @tdas for spending time on this PR. Closing this as you suggested.
